### ISSUE TITLE
Dynamic silently refuses to load over-long salts, but would ideally warn

### ISSUE THREAD

**Issue by rodrigozanatta** [openwall/john#5009]
I found a bug if I try to use a dynamic function with salt bigger than 32 character.

**john -test -format:"dynamic=sha512($u.$p.$s)"**

```
echo wordlist.txt >
password
```

```
echo > test.in
User1:145605ea09b770ffae598cdb1a1828dc6c66801c19716bbd59f64bfbb2a06e13ad72fb4896d6b180e1c297fd49680014fac80616f56e20d73f9350f150ff6a1f$NormalSalt
User2:c8620f4f0001dc11e61c134f541fd485cd875d77ab8c3f6112f1d821c2fea36018b32c6db96490cd321d4361b299d79f1cda27d7d2ad40e15d4437bca2e9acd2$SaltWithLimitOf32characterWorkss
Unbreakable:d8628ba8d3b35c21f5801216545bb232ff0280d04eb6b7b552217b1da117762507f457f5629216d0a6576efefb4df57de7a1acd8f28b6fc2f017a2d2aa991459$SaltWithMoreThan32characterWillNotWork
```

So, if I run the program:
**john  --format:"dynamic=sha512($u.$p.$s)" --wordlist=wordlist.txt  test.in**
Only found two:
```
password         (User1)
password         (User2)
```

To solve this, I can just change the order of the user and salt

**john -test -format:"dynamic=sha512($s.$p.$u)"**

```
echo > solved.in
NormalSalt:145605ea09b770ffae598cdb1a1828dc6c66801c19716bbd59f64bfbb2a06e13ad72fb4896d6b180e1c297fd49680014fac80616f56e20d73f9350f150ff6a1f$User1
SaltWithLimitOf32characterWorkss:c8620f4f0001dc11e61c134f541fd485cd875d77ab8c3f6112f1d821c2fea36018b32c6db96490cd321d4361b299d79f1cda27d7d2ad40e15d4437bca2e9acd2$User2
SaltWithMoreThan32characterWillNotWork:d8628ba8d3b35c21f5801216545bb232ff0280d04eb6b7b552217b1da117762507f457f5629216d0a6576efefb4df57de7a1acd8f28b6fc2f017a2d2aa991459$Unbreakable
```

**john  --format:"dynamic=sha512($s.$p.$u)" --wordlist=wordlist.txt  solved.in**
Now every password is found:
```
password         (NormalSalt)
password         (SaltWithLimitOf32characterWorkss)
password         (SaltWithMoreThan32characterWillNotWork)
```
This is a VERY difficult bug to find! And there are some problems with it. 

I have a bad salt (with new line). So I input it in hex. But...

- Because the salt is bigger than 32, I can't input it.
- If I use the hex salt as User, the John don't accept.

So, how can I solve it? How can I input the **User** as **hex**? Only after the bug is solved?

This is my version (the last):
```
John the Ripper 1.9.0-jumbo-1 OMP [cygwin 64-bit x86_64 AVX2 AC]
Copyright (c) 1996-2019 by Solar Designer and others
```

**Comment by magnumripper** [comment]
A limitation of salt length is hardly what I'd call a bug but that said, it'd be nice to fix it.

The problem is related to optimizations: For maximum speed, SIMD formats can only have a total input of 55 bytes - sum of salts and password lengths becuase that's a single Merkel-Damgard block. I believe most/all dynamic formats extend it to the next block limit: 64 + 55 so 119 bytes. A problem with SIMD is that if running eg. AVX2, we do 8 candidates in parallel for more or less 8x the speed. Now if one of them gets larger than 119 bytes, the other seven will need special treatment and in the end it will be as slow as if all eight of them broke that length threshold.

**Comment by solardiz** [comment]
I assume the bug here is those salts are not rejected right away. Of course, ideally they'd be supported, but that isn't essential to fixing the bug aspect of this issue.

Then there's the question. I _guess_ building with `--disable-simd` _might_ work around the issue, but I am not sure.

**Comment by solardiz** [comment]
@magnumripper If it's about fitting in the underlying hash's block, then why does re-ordering of salt and username matter? Maybe it's something else this time.

**Comment by solardiz** [comment]
Oh, actually `john` _does_ reject _loading_ of the hash with the over-long salt. Trying `test.in` above, with 3 lines in it, I get `Loaded 2 password hashes with 2 different salts (dynamic=sha512($u.$p.$s) [128/128 AVX 2x])`. So no bug here, after all, and I was too quick to comment.

However, we could want to enhance the loader to print warnings about unsupported hashes, which could be helpful especially for new users. Maybe enabling some "expert mode" setting should disable that. We can open a separate issue for that enhancement.

We already print warnings suggesting other formats when running without `--format`. Historically, `--format` meant "I know what I am doing, and yes I want anything unsupported or of other formats to be silently skipped". However, maybe that should change.

**Comment by solardiz** [comment]
> we could want to enhance the loader to print warnings about unsupported hashes

Actually, for this specific issue we'd probably need to add such printing to dynamic's `valid()`, like we already have in some other formats that can sometimes reject an almost-valid-but-not-quite-supported hash.

**Comment by rodrigozanatta** [comment]
> Oh, actually john does reject loading of the hash with the over-long salt. Trying test.in above, with 3 lines in it, I get Loaded 2 password hashes with 2 different salts (dynamic=sha512($u.$p.$s) [128/128 AVX 2x]). So no bug here, after all, and I was too quick to comment.

Lol, sorry for this confusion. **The problem is that John don't LOAD the hash.** I should have to say it before. 

So I thought it was just a format error. Because... after all, the program find the same password.
I just don't understand this:
- The variable $u accept long string
- The variable $s don't

In the end, this is just random string in both sides
sha512($u.$p.$s) == sha512($s.$p.$u)

The program CAN calculate any random string inside the sha512 function. Internally, what difference the program does when I change $s with $u? 

Even if it doesn't run in the best optimization, certainly it will run better than my own code.
